REGRESSION (Sonoma?): RTL scrollbar incorrect spacing

nmoucht opened the pull request:
#### e7a46d11ffb4d88a4075971be726d8999d8e4230
<pre>
REGRESSION (Sonoma?): RTL scrollbar incorrect spacing
<a href="https://bugs.webkit.org/show_bug.cgi?id=263512">https://bugs.webkit.org/show_bug.cgi?id=263512</a>
<a href="https://rdar.apple.com/117355120">rdar://117355120</a>

Reviewed by Simon Fraser.

We need to pass changes to shouldPlaceVerticalScrollbarOnLeft to the UI-process
so that we can properly set userInterfaceLayoutDirection on the scroller imp.
Since on first style change the scrollable area in styleDidChange() doesn&apos;t yet
have a render layer backing, also pass accross UserInterfaceLayoutDirection
when we first set the scrollbar layer as well.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::shouldPlaceVerticalScrollbarOnLeft const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::RemoteScrollbarsController):
(WebKit::RemoteScrollbarsController::mouseMovedInContentArea):

Canonical link: <a href="https://commits.webkit.org/276040@main">https://commits.webkit.org/276040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcd969a3b9eb74ca01396810f5d255b43496bf5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43517 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46152 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39646 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35982 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37492 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16956 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38556 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1573 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39695 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47698 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18547 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15179 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42768 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41441 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9705 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20149 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19600 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
```diff
@@ -60,6 +60,7 @@ ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode(
     MouseLocationState&& mouseLocationState,
     ScrollbarHoverState&& scrollbarHoverState,
     ScrollbarEnabledState&& scrollbarEnabledState,
+    bool shouldPlaceVerticalScrollbarOnLeft,
```
I think we could use the `UserInterfaceLayoutDirection` enum everywhere, and call it "scrollbarLayoutDirection".

nmoucht replied:
Fixed.

cdumez commented on LayoutTests/fast/scrolling/mac/scrollbars/rtl-scrollbar-state.html:
```diff
@@ -0,0 +1,65 @@
+            height: 200%;
+        }
+    </style>
+    <script src="../../../../resources/js-test-pre.js"></script>
```
Would be nice to use `js-test.js` in new tests instead of the legacy `js-test-pre.js` / `js-test-post.js`.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -373,6 +374,21 @@ void AsyncScrollingCoordinator::stopAnimatedScroll(ScrollableArea& scrollableAre
+    if (!scrollingNodeID)
+        return;
+
+    auto stateNode = dynamicDowncast<ScrollingStateScrollingNode>(m_scrollingStateTree->stateNodeForID(scrollingNodeID));
```
`auto` -> `RefPtr`

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -89,6 +89,12 @@
 
+    if (scrollingStateNode.hasChangedProperty(ScrollingStateNode::Property::ShouldPlaceVerticalScrollbarOnLeft)) {
+        auto shouldPlaceVerticalScrollbarOnLeft = scrollingStateNode.shouldPlaceVerticalScrollbarOnLeft();
+        m_scrollerPair->horizontalScroller().setShouldPlaceVerticalScrollbarOnLeft(shouldPlaceVerticalScrollbarOnLeft);
```
I am no area expert. I would just like to comment to say that I find it a little confusing to be setting a "Vertical Scrollbar" property on a *horizontal* scroller :)

nmoucht replied:
Haha yeah changed to using UserInterfaceLayoutDirection to make this more clear.

cdumez commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
```diff
@@ -40,6 +40,12 @@
 
+void RemoteScrollbarsController::rtlStyleDidChange(bool shouldPlaceVerticalScrollbarOnLeft)
+{
+    if (auto scrollingCoordinator = m_coordinator.get())
```
`auto` -> `RefPtr`. I assume it already is but we shouldn't hide it.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -73,12 +73,12 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
 
     virtual void hasNodeWithAnimatedScrollChanged(bool) { };
     
+    WEBCORE_EXPORT void setUserInterfaceLayoutDirection(ScrollableArea&, UserInterfaceLayoutDirection) override;
```
I think this should be called `scrollbarLayoutDirection` everywhere; it's less general that "user interface layout direction".

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinator.h:
```diff
@@ -33,6 +33,7 @@
 #include "ScrollSnapOffsetsInfo.h"
 #include "ScrollTypes.h"
 #include "ScrollingCoordinatorTypes.h"
+#include "UserInterfaceLayoutDirection.h"
```
You should be able to forward-declare the enum.

nmoucht replied:
I couldn't find a good way to do this since it's needed on Page.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -89,6 +89,12 @@
 
+    if (scrollingStateNode.hasChangedProperty(ScrollingStateNode::Property::ScrollbarLayoutDirection)) {
+        auto scrollbarLayoutDirection = scrollingStateNode.scrollbarLayoutDirection();
+        m_scrollerPair->horizontalScroller().setScrollbarLayoutDirection(scrollbarLayoutDirection);
```
Does this do anything?

nmoucht replied:
We set this on both scroller imps with ui-side compositing off but yeah I don't think this is strictly necessary.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -345,6 +346,11 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
 
+    if (layer() && oldStyle && oldStyle->shouldPlaceVerticalScrollbarOnLeft() != newStyle.shouldPlaceVerticalScrollbarOnLeft()) {
+        if (auto* scrollableArea = layer()->scrollableArea())
+            scrollableArea->scrollbarsController().rtlStyleDidChange(shouldPlaceVerticalScrollbarOnLeft() ? UserInterfaceLayoutDirection::RTL : UserInterfaceLayoutDirection::LTR);
```
`rtlStyleDidChange` would be better named as `scrollbarLayoutDirectionChanged`.

nmoucht replied:
Fixed.

nmoucht committed:
REGRESSION (Sonoma?): RTL scrollbar incorrect spacing
https://bugs.webkit.org/show_bug.cgi?id=263512
rdar://117355120

Reviewed by Simon Fraser.

We need to pass changes to shouldPlaceVerticalScrollbarOnLeft to the UI-process
so that we can properly set userInterfaceLayoutDirection on the scroller imp.
Since on first style change the scrollable area in styleDidChange() doesn't yet
have a render layer backing, also pass accross UserInterfaceLayoutDirection
when we first set the scrollbar layer as well.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::shouldPlaceVerticalScrollbarOnLeft const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::setShouldPlaceVerticalScrollbarOnLeft):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::RemoteScrollbarsController):
(WebKit::RemoteScrollbarsController::mouseMovedInContentArea):

Canonical link: https://commits.webkit.org/276040@main
